### DISCUSSION TITLE
Storybook doesn't work in monorepo - vue 3 vite 3

### DISCUSSION THREAD

**Discussion by singhinderpal** [storybookjs/builder-vite#477] [Q&A]
Hi there,

I am trying to setup storybook in my vue 3 vite 3 package inside of a monorepo.
The steps I followed:
Navigate to `monorepo\packages\vue3project` and follow the following commands:
```
pnpx storybook init -s
pnpm i
pnpm i -D @storybook/cli
pnpm storybook
```

I get the following error when I try to open locally started storybook from `http://localhost:6006/`

```
[vite] Internal server error: Failed to resolve import "@storybook/preview-web" from "..\..\..\..\..\virtual:\@storybook\builder-vite\vite-app.js". Does the file exist?
  Plugin: vite:import-analysis
  File: /virtual:/@storybook/builder-vite/vite-app.js
  1  |  import { composeConfigs, PreviewWeb } from '@storybook/preview-web';
     |                                              ^
  2  |      import { ClientApi } from '@storybook/client-api';
  3  |      import '/virtual:/@storybook/builder-vite/setup-addons.js';
```

I then crosschecked the nodemodules and it does not contain a `@storybook/preview-web`.
![image](https://user-images.githubusercontent.com/62807061/183281954-248cee35-d90e-4d48-bf0c-775d0f91deb7.png)

That's my main.cjs
```
module.exports = {
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions"
  ],
  "framework": "@storybook/vue3",
  "core": {
    "builder": "@storybook/builder-vite"
  },
  "features": {
    "storyStoreV7": true
  }
}
```

**Comment by fivaz** [comment]
The problem is with pnp not vite3 or mono repo. Try the same using npm or yarn v1 and it will work fine. 

If you want a quickly but wrong fix, download all those dependencies as dev dependencies of your own project.

**Reply by singhinderpal** [reply]
I am using pnpm!
The problem is that if I run `pnpx storybook init` without `-s` flag, it installs all node modules at the package level which aren't deep linked from mono repo and a package-lock.json file! (just like simple `npx storybook init` from `npm` would)

This means that a new dev cloning this for first time will have to do `pnpm i` at the root mono repo level and `npm i` at the package level, which is not good!

If I use it with -s flag, I get the above mentioned error!

**Reply by singhinderpal** [reply]
I also tried adding failing dependencies manually as dev dependencies and I added approx 9 dependencies `@storybook/something` after which I left hope for this being a quick fix.

**Reply by IanVS** [reply]
Unfortunately there are about 20 dependencies you'd need to add.  The current workarounds are to add them, or to use the shameful hoisting option from pnpm.  There's an issue tracking this at https://github.com/storybookjs/builder-vite/issues/55.  

There's a chance that this situation will be fixed or at the least improved in the upcoming storybook version 7.  But that is still at least a few months away.

**Reply by singhinderpal** [reply]
@IanVS If you don't mind, can you please share the list of 20 deps that you mentioned, that will be a big help for me!
I currently have following `@storybook/addon-docs @storybook/addon-backgrounds @storybook/client-api @storybook/preview-web @storybook/addon-measure @storybook/addon-outline @storybook/channel-postmessage @storybook/channel-websocket @storybook/addons`

**Reply by IanVS** [reply]
I don't keep an up-to-date list, and it might have changed, but here's what someone used to do the same thing: https://github.com/jdk2pq/vite-storybook-pnpm-vue/commit/9ccdf744827b7d539440849f545df50c5675c981, it might be a useful starting point.

**Reply by singhinderpal** [reply]
@IanVS thanks! 
I will try this and will post a comment here if I find any differences for future devs if something else needs to add up.

**Comment by singhinderpal** [comment]
@IanVS 
After the default `pnpx storybook init -s`, I had to add the following dependencies in total!
At first glance, it seems working fine!

```
@babel/preset-env
@mdx-js/react
@storybook/addon-backgrounds
@storybook/addon-docs
@storybook/addon-measure
@storybook/addon-outline
@storybook/addons
@storybook/channel-postmessage
@storybook/channel-websocket
@storybook/client-api
@storybook/client-logger
@storybook/core-common
@storybook/preview-web
@vue/compiler-sfc
react
react-dom
```

**Reply by IanVS** [reply]
By the way, this should no longer be needed with Storybook 7.0.  pnpm support has been greatly improved!

**Comment by geoff-harper** [comment]
I've tried about every permutation of `pnpx` command at varying levels of my monorepo and I haven't really gotten anywhere (beta 46). Is there a step by step recipe for setting up v7 beta in the following conditions?

- pnpm monorepo
- vite
- react
- storybook down in `packages/*`

It seems most attemps end with https://github.com/storybookjs/storybook/issues/12995#issuecomment-748032086. Even if it's from a fresh start no migration, I'd love to get this up and running

**Reply by IanVS** [reply]
What problem are you hitting?

**Reply by geoff-harper** [reply]
In spectacular fashion I tried it again fresh off the weekend with coffee in hand and it seems to be working as expected. For anyone having a similar issue

- cd down into your ui package
- upgrade deps (I did this a few days ago can't remember specifics, but the storybook dev dep is in ui's package file, not root)
- remove contents of `.storybook` dir
- `pnpx sb@next init`

Not sure which parts of this are fluff or required but this is how I've got it going. Thanks for the prompt reply, appreciate the work
